### PR TITLE
[SR] Support replays for crashes in buffer and session modes

### PR DESCRIPTION
**Pull request by romtsn**
_#skip-changelog_

## :scroll: Description

* Make buffer mode work for ANRs
  * For that we have to store errorSampleRate to disk and then roll the dice on next launch, because it could change in the meantime
  * Also, since in buffer mode we don't actually write `replayId` to scope until an error happens (and in the case of an ANR we don't know until next restart), the persisted `replayId` to disk might be irrelevant. So we have to iterate through the leftover  replay folders and find the one that was last modified closest to the ANR and take it as the previous replay
  * Important part is: we set the relevant `replayId` back to the disk cache, so later `ReplayIntegration` can pick it up and finalize and send it. The order will be ensured by an integration test
* Make buffer mode work for Crashes
  * Renamed `sendReplayForEvent` to `captureReplay` which just accept one argument now to define whether the app process is terminating or not. If it is, then we don't capture the buffered replay in the current process, but will capture it on next launch, to avoid weird issues with half-captured videos
  * Refactored `createSegment` method and moved it to a static method under `CaptureStrategy` so we could call it from `ReplayIntegration` as well
* Small fix for network breadcrumbs where we could deserialize double values sometimes (didn't want to change the deserializer for that to not break something accidentally)
* Small fix for the video encoder, set IFRAME_INTERVAL to -1 so it always only encode partial updates without full frames to save the video size

## :bulb: Motivation and Context

Part of https://github.com/getsentry/sentry/issues/74441

## :green_heart: How did you test it?
manually + automated

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/AnrV2EventProcessor.java [hunk 1]
```diff
@@ -156,14 +170,67 @@ private void backfillScope(final @NotNull SentryEvent event, final @NotNull Obje
...
+    final @Nullable String replayErrorSampleRate =
+        PersistingOptionsObserver.read(options, REPLAY_ERROR_SAMPLE_RATE_FILENAME, String.class);
+
+    if (replayErrorSampleRate == null) {
+      return false;
+    }
+
+    try {
+      // we have to sample here with the old sample rate, because it may change between app launches
+      final SecureRandom random = this.random != null ? this.random : new SecureRandom();
```
Comment (markushi) [inline sentry-android-core/src/main/java/io/sentry/android/core/AnrV2EventProcessor.java:183]:
```suggestion
      final @NotNull SecureRandom random = this.random != null ? this.random : new SecureRandom();
```

File: sentry-android-core/src/main/java/io/sentry/android/core/AnrV2EventProcessor.java [hunk 2]
```diff
@@ -156,14 +170,67 @@ private void backfillScope(final @NotNull SentryEvent event, final @NotNull Obje
...
+
+    return true;
+  }
+
   private void setReplayId(final @NotNull SentryEvent event) {
-    final @Nullable String persistedReplayId =
-        PersistingScopeObserver.read(options, REPLAY_FILENAME, String.class);
+    @Nullable
+    String persistedReplayId = PersistingScopeObserver.read(options, REPLAY_FILENAME, String.class);
+    final File replayFolder = new File(options.getCacheDirPath(), "replay_" + persistedReplayId);
```
Comment (markushi) [inline sentry-android-core/src/main/java/io/sentry/android/core/AnrV2EventProcessor.java:205]:
```suggestion
    final @NotNull File replayFolder = new File(options.getCacheDirPath(), "replay_" + persistedReplayId);
```

File: sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt [hunk 3]
```diff
@@ -259,4 +256,67 @@ public class ReplayIntegration(
...
+        // clean up old replays
+        options.cacheDirPath?.let { cacheDir ->
+            File(cacheDir).listFiles { dir, name ->
+                if (name.startsWith("replay_") &&
+                    !name.contains(replayId.toString()) &&
+                    !(unfinishedReplayId.isNotBlank() && name.contains(unfinishedReplayId))
+                ) {
+                    FileUtils.deleteRecursively(File(dir, name))
+                }
+                false
```
Comment (markushi) [inline sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt:270]: I don't think the `.listFiles(filter: FileNameFilter)` ([docs here](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#listFiles--)) is meant to be used this way, as your filter code deletes folders while it is still iterating 😅 How about this?
```suggestion
            File(cacheDir).listFiles()?.forEach { file ->
                val name = file.name
                if (name.startsWith("replay_") &&
                    !name.contains(replayId.toString()) &&
                    !(unfinishedReplayId.isNotBlank() && name.contains(unfinishedReplayId))
                ) {
                    FileUtils.deleteRecursively(file)
                }
            }
```

File: sentry-android-replay/src/main/java/io/sentry/android/replay/capture/BufferCaptureStrategy.kt [hunk 4]
```diff
@@ -228,7 +179,75 @@ internal class BufferCaptureStrategy(
...
+            options.logger.log(ERROR, e, "Failed to delete replay segment: %s", file.absolutePath)
+        }
+    }
+
+    private fun MutableList<ReplaySegment.Created>.capture() {
+        var bufferedSegment = removeFirstOrNull()
+        while (bufferedSegment != null) {
+            bufferedSegment.capture(hub)
+            bufferedSegment = removeFirstOrNull()
+            Thread.sleep(100L)
```
Comment (markushi) [inline sentry-android-replay/src/main/java/io/sentry/android/replay/capture/BufferCaptureStrategy.kt:204]: reserved for future performance optimization? 😄 In any case probably makes sense to add a comment why you're doing that.

File: sentry-android-replay/src/main/java/io/sentry/android/replay/capture/CaptureStrategy.kt [hunk 5]
```diff
@@ -34,4 +51,190 @@ internal interface CaptureStrategy {
...
+            frameRate: Int,
+            videoDuration: Long,
+            replayType: ReplayType,
+            screenAtStart: String?,
+            breadcrumbs: List<Breadcrumb>,
+            events: LinkedList<RRWebEvent>
+        ): ReplaySegment {
+            val endTimestamp = DateUtils.getDateTime(segmentTimestamp.time + videoDuration)
+            val replay = SentryReplayEvent().apply {
+                eventId = currentReplayId
```
Comment (markushi) [inline sentry-android-replay/src/main/java/io/sentry/android/replay/capture/CaptureStrategy.kt:130]: nit, for readability I'd use `this.` everywhere

**Review by markushi — APPROVED** [review]
Nice one, great to see this all coming together. Left a few minor comments

File: sentry-android-replay/src/main/java/io/sentry/android/replay/capture/BufferCaptureStrategy.kt (hunk 4 above)
Comment (romtsn) [reply sentry-android-replay/src/main/java/io/sentry/android/replay/capture/BufferCaptureStrategy.kt:204]: it's rather not to burst the server with envelopes and active rate limiting, similar to what we do here: https://github.com/getsentry/sentry-java/blob/3a89243fa0ec48fc16193070c2224833d94f9156/sentry/src/main/java/io/sentry/DirectoryProcessor.java#L103-L106

but great point about adding the comment, will do 👍

File: sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt (hunk 3 above)
Comment (romtsn) [reply sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt:270]: yeah, I've figured this in some other place but forgot to change here 😅 thanks for pointing out!

**Pull request merged**